File: js/examples/k-means/fan.js
"use strict";

let zmq = require("zeromq");
process.stdin.resume();
process.stdin.setRawMode(true);

let sender = zmq.socket("push");
sender.bind("tcp://*:5557");

let sink = zmq.socket("push");
sink.bind("tcp://localhost:5558");

let rep = zmq.socket("rep");
rep.bind("tcp://localhost:5559");

let data = [
	{ x: 0, y: 0 },
	{ x: 2, y: 0 },
	{ x: 1, y: 0 },
	{ x: 3, y: 0 },
	{ x: 4, y: 0 },
	{ x: 0, y: 5 },
	{ x: 0, y: 2 },
	{ x: 0, y: 1 },
	{ x: 1, y: 3 },
	{ x: 2, y: 1 },
	{ x: 3, y: 2 },
	{ x: 4, y: 4 },
];

let u = [
	{ x: 1, y: 1 },
	{ x: 2, y: 0 },
];

function formatMessage(points) {
	let obj = {
		centroids: u,
		points: points,
	};
	return JSON.stringify(obj);
}

function work() {
	let start = 0,
		end = 0;
	while (start < data.lenght) {
		end = Math.min(data.lenght, end + sz);
		let msg = formatMessage(data.slice(start, end));
		sender.send(msg);
		start = end;
	}
}

rep.on("message", (data) => {
	let sz = 4;
	let request = JSON.parse(data);
	if (request.type == "next") {
		work();
	}
	if (request.type == "finish") {
		console.log("yeah :)");
	}
});

console.log("Press enter when the workers are ready…");
process.stdin.on("data", function () {
	process.stdin.pause();
	work();
});

File: challenges/six-challenge/fan.py
import zmq
import random
import time
import json
import random
import math

try:
    raw_input
except NameError:
    # Python 3
    raw_input = input

context = zmq.Context()

# Socket to send messages on
sender = context.socket(zmq.PUSH)
sender.bind("tcp://*:5557")

# Socket with direct access to the sink: used to synchronize start of batch
sink = context.socket(zmq.PUSH)
sink.connect("tcp://localhost:5558")

sinkRep = context.socket(zmq.REP)
sinkRep.bind("tcp://*:5559")

print("Input the number of centroids  when the workers are ready: ")
k = raw_input()
print("Sending tasks to workers…")

numberPoints = 150
segment = 3
dimensions = 4

# The first message is "0" and signals start of batch
sink.send_string(json.dumps([k,dimensions, numberPoints]))


def generateCentroid(start, end, size):
	return [random.uniform(start, end) for i in range(size)]

def readFile(nameFile):
	with open(nameFile) as file:
		for i,line in enumerate(file):
			print([float(x) for x in line.split(",")])

def mapper(value, maximum, lenRange):
	return [value, min(maximum, value + lenRange)]

def work(sizeData, numberIntervals, centroids):
	lenRange = math.ceil(sizeData / numberIntervals)
	rangeWork = [mapper(x,sizeData,lenRange) for x in range(0,sizeData,lenRange)]
	for work in rangeWork:
		task = { "centroids": centroids, "range": work}
		sender.send_string(json.dumps(task))

work(numberPoints,12, [ generateCentroid(-3 , 3, dimensions) for i in range(int(k))])


numberIterations = 0

while numberIterations < 5:
	request = sinkRep.recv()
	centroids = json.loads(request)
	numberIterations += 1
	sinkRep.send_string('ok')
	work(numberPoints,12,centroids)